lab2/part1/myrouter_part1.py

```diff
@@ -0,0 +1,149 @@
+'''
+class ARPContext:
+    def __init__(self):
+        self.map = {}
+        return
+    
+    def __iter__(self):
+        return iter(self.map)
+
+    def __getitem__(self, item):
+        if item in self.map and isinstance(self.map[item], ARPContext.ARPEntry):
+            return self.map[item].mac_addr
+        else:
+            return None
+
+    def __setitem__(self, ip_addr, mac_addr):
+        self.add_mapping(ip_addr, mac_addr)
+        return
+
+    #Add or updates an IP to MAC address mapping
+    #  ip_addr      IPv4Address
+    #  mac_addr     EthAddr
+    def add_mapping(self, ip_addr, mac_addr):
+        if isinstance(ip_addr, IPv4Address) and isinstance(mac_addr, EthAddr):
+            self.map[ip_addr] = ARPContext.ARPEntry(mac_addr)
+        return
+
+    #Process an ARP requests in the provided packet. Return a reply packet
+    #if the target IP address is in our table
+    #   pkt         Arp packet to process
+    def arp_request(self, pkt: Arp):
+        reply_ip  = pkt.targetprotoaddr
+        reply_mac = self.map[reply_ip].mac_addr
+
+        #Construct reply packet
+        etp = Ethernet(ethertype = EtherType.ARP,
+                       src       = reply_mac,
+                       dst       = pkt.senderhwaddr)
+
+        arp = Arp(operation         = ArpOperation.Reply,
+                  senderhwaddr      = reply_mac,
+                  senderprotoaddr   = reply_ip,
+                  targethwaddr      = pkt.senderhwaddr,
+                  targetprotoaddr   = pkt.senderprotoaddr)
+
+        return_pkt = etp + arp
+        return return_pkt
+
+    #Process an ARP reply in the provided packet.
+    #   pkt         Arp packet to process
+    def arp_reply(self, pkt: Arp):
+        self.add_mapping(pkt.senderprotoaddr, pkt.senderhwaddr)
+
+    """
+    Class:          ARPContext.ArpEntry
+    Description:    Table entry pairing a timestamp with a MAC address
+    """
+    class ARPEntry:
+        def __init__(self, mac_addr, time=time.time()):
+            self.mac_addr   = mac_addr
+            self.timestamp  = time
+            return
+
+# end class ARPContext
+
+class Router(object):
+    def __init__(self, net):
+        self.net = net
+        self.local_proto_eth = ARPContext()     #Local address maps
+        self.other_proto_eth = ARPContext()     #Other address maps
+
+        #Cache IP->MAC mapping for local interfaces
+        for intf in net.interfaces():
+            self.local_proto_eth[intf.ipaddr] = intf.ethaddr
+        
+    #Main Router loop
+    def router_main(self):    
+        '''
+        Main method for router; we stay in a loop in this method, receiving
+        packets until the end of time.
+        '''
+        while True:
+            gotpkt = True
+            try:
+                _, input_port, pkt = self.net.recv_packet(timeout=1.0)
+            except NoPackets:
+                log_debug("No packets available in recv_packet")
+                gotpkt = False
```

Comment:
Replacing this line with a `continue` here removes the need for the variable `gotpkt`, and IMO improves readability.